lib/spack/spack/test/git_fetch.py

```diff
@@ -251,6 +251,41 @@ def test_gitsubmodule(submodules, mock_git_repository, config,
                 assert not os.path.isfile(file_path)
 
 
+@pytest.mark.disable_clean_stage_check
+@pytest.mark.parametrize("submodules", [list, str, None])
+def test_gitsubmodules_callable(
+    submodules, mock_git_repository, config, mutable_mock_repo, monkeypatch
+):
+    """
+    Test GitFetchStrategy behavior with submodules selected after concretization
+    """
+    def submodules_callback(package):
+        name = 'third_party/submodule0'
+        return [name] if submodules is list else name if submodules is str else None
```

Comment:
As commented above, I would avoid the string and the `None` cases to keep the implementation simpler. Also, not sure how easy it is, but it would be better for readability if all the setup is in one or more fixtures, so it is clear what the test is asserting.